File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.go
func dfs(manager int, adj map[int][] int, informTime []int, count int, result *int){
    
    if count > *result {
        *result = count
    }
    
    for _, employee := range adj[manager] {
        
        dfs(employee, adj, informTime, count + informTime[manager], result)
    }
}

func numOfMinutes(n int, headID int, manager []int, informTime []int) int {
    
    adj := map[int][]int{}
    
    for i:=0; i<n; i++ {
        
        managerID := manager[i]
        employeeID := i
        
        adj[managerID] = append(adj[managerID], employeeID)
    }
    
    result := 0
    count := 0
    
    dfs(headID, adj, informTime, count, &result);
    
    return result
}

File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.java
class Solution {
    public void dfs(int manager, List<List<Integer>> adj, int[] informTime, int curTime, int[] result){
        
        result[0] = Math.max(result[0], curTime);
        
        for(int employee : adj.get(manager)){
            
            dfs(employee, adj, informTime, curTime + informTime[manager] , result);
        }
    }
    
    public int numOfMinutes(int n, int headID, int[] manager, int[] informTime) {
        
        List<List<Integer>> adj = new ArrayList<>(n);
        for(int i=0; i<n; i++){
            adj.add(new ArrayList<>());
        }
        
        for(int i=0; i<n; i++){
            
            int managerID = manager[i];
            int employeeID = i;
            
            if(managerID == -1)
                continue;
            
            adj.get(managerID).add(employeeID);
        }
        
        int[] result = new int[1];
        
        dfs(headID, adj, informTime, 0, result);
        
        return result[0];
    }
}